src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/vmware/VMWareProviderVirtualMachineUtil.java

```diff
@@ -84,9 +72,110 @@ public VirtualMachine searchVirtualMachineByName(String name, Folder rootFolder)
+        return (HostSystem) new InventoryNavigator(rootFolder).searchManagedEntity("HostSystem", name);
+    }
+
+    public Folder searchVMFolderFromVMName(String name, Folder rootFolder) throws RemoteException {
+        ManagedEntity current = searchVirtualMachineByName(name, rootFolder).getParent();
+        while (!(current instanceof Folder)) {
+            current = current.getParent();
+        }
+        return (Folder) current;
+    }
+
+    public Folder searchVMFolderByHostname(String name, Folder rootFolder) throws RemoteException {
+        return Lists.newArrayList(new InventoryNavigator(rootFolder).searchManagedEntities("Folder"))
+                    .stream()
+                    .map(folder -> (Folder) folder)
+                    .filter(folder -> folder.getName().toLowerCase().contains("vm"))
+                    .filter(folder -> folder.getName().toLowerCase().contains(name.toLowerCase()))
+                    .findAny()
+                    .orElse(null);
+    }
+
+    public ResourcePool searchResourcePoolByHostname(String hostname, Folder rootFolder) throws RemoteException {
+        return Lists.newArrayList(new InventoryNavigator(rootFolder).searchManagedEntities("ResourcePool"))
+                    .stream()
+                    .map(resourcePool -> (ResourcePool) resourcePool)
+                    .filter(resourcePool -> resourcePool.getParent().getName().equals(hostname))
+                    .findAny()
+                    .orElse(null);
+    }
+
+    public Datastore getDatastoreWithMostSpaceFromHost(HostSystem host) throws RemoteException {
+        return Lists.newArrayList(host.getDatastores())
+                    .stream()
+                    .filter(datastore -> datastore.getSummary().isAccessible())
+                    .max(Comparator.comparing(datastore -> (int) datastore.getSummary().getFreeSpace()))
+                    .orElseGet(null);
+    }
+
+    public Datastore getDatastoreWithMostSpaceFromPool(ResourcePool resourcePool) throws RemoteException {
+        return Lists.newArrayList(resourcePool.getOwner().getDatastores())
+                    .stream()
+                    .filter(datastore -> datastore.getSummary().isAccessible())
+                    .max(Comparator.comparing(datastore -> (int) datastore.getSummary().getFreeSpace()))
+                    .orElseGet(null);
+    }
+
+    public ResourcePool getRandomResourcePool(Folder rootFolder) throws RemoteException {
+        List<ResourcePool> resourcePools = Lists.newArrayList(new InventoryNavigator(rootFolder).searchManagedEntities("ResourcePool"))
+                                                .stream()
+                                                .map(resourcePool -> (ResourcePool) resourcePool)
+                                                .collect(Collectors.toCollection(ArrayList::new));
+        return resourcePools.isEmpty() ? null : resourcePools.get(new Random().nextInt(resourcePools.size()));
```

Comment:
You return null instead of an empty list? What additional information gives you null compared to an empty list? 
Null forces you to work with Optional or do null checks. Ensuring to never return makes the code cleaner. Except if you really need the additional information which returning null gives you. 
But in this case you seem to set it equal to an empty list.